README.md
## Multi-Checkbox Web Component
[![Published on npm](https://img.shields.io/badge/npm-published-blue)](https://www.npmjs.com/package/multi-checkbox)
[![Version](https://img.shields.io/badge/version-0.1.11-brightgreen)](https://github.com/smontanus/multi-checkbox)

Multi-Checkbox is a web component that utilizes multiple checkboxes in a dropdown list to populate the value of an input. It is written in pure Javascript with no dependencies.

### Installation
* **NPM**
    `npm install multi-checkbox`

* **CDN**
    ```<script type="module" src="https://www.unpkg.com/multi-checkbox/multi-checkbox.js"></script>```

### Usage
The Multi-Checkbox component utilizes the custom element tag `<multi-checkbox>`. Within the custom tag the component requires a slotted `<ul>` element with the slot attribute set to `check-values`. Within the slotted `<ul>` each `<li>` element represents a checkbox label and value for the component. The custom element has two attributes:

* separator - The separator string used between the checked values.
* value - The custom element component value.

The component will automatically update when attributes are set dynamically. So changing the separator attribute dynamically will automatically replace the value attribute with the updated separator. Changing the value attribute dynamically will automatically check the corresponding checkboxes, if they exist. Additionally, if the `<li>` elements are changed dynamically the element will reset with the new checkbox values.

```html
<multi-checkbox separator="," value="">
    <ul slot="check-values">
        <li>Item 1</li>
        <li>Item 2</li>
        <li>Item 3</li>
    </ul>
</multi-checkbox>
```

### CSS Style Variables
The following CSS property variables, along with their default values, are available for custom styling.
|Property Variable|Default Value|
|---|---|
|***General Element***|
|--mc-border|1px solid #000000|
|--mc-border-radius|2px|
|--mc-display|inline-block|
|--mc-font|400 0.9em Arial|
|--mc-margin|0|
|--mc-vertical-align|middle|
|--mc-z-index|0|
|***Dropdown***|
|--mc-dropdown-background|#efefef|
|--mc-dropdown-box-shadow|3px 3px 5px 1px rgba(0,0,0,0.35)|
|--mc-dropdown-max-height|100px|
|--mc-dropdown-text-align|left|
|--mc-dropdown-width|209px|
|***Input***|
|--mc-target-background|#efefef|
|--mc-target-color|#000000|
|--mc-target-cursor|default|
|--mc-target-height|28px|
|--mc-target-margin|0|
|--mc-target-max-height|default|
|--mc-target-min-height|default|
|--mc-target-max-width|default|
|--mc-target-min-width|default|
|--mc-target-padding|0px 0px 0px 3px|
|--mc-target-outline|none|
|--mc-vertical-align|middle|
|--mc-target-width|175px|
|***Dropdown Line Items***|
|--mc-dropdown-color|#000000|
|--mc-dropdown-line-height|2em|
|--mc-ul-margin|5px|
|--mc-ul-padding|0|
|--mc-checkbox-height|auto|
|--mc-checkbox-width|auto|
|***Toggle Button***|
|--mc-toggle-button-background|#ffffff|
|--mc-toggle-button-color|#000000|
|--mc-toggle-button-cursor|pointer|
|--mc-toggle-button-height|30px|
|--mc-toggle-button-outline|none|
|--mc-vertical-align|middle|
|--mc-toggle-button-width|30px|
|***Close Button***|
|--mc-close-button-background|#efefef|
|--mc-close-button-border|none|
|--mc-close-button-border-radius|default|
|--mc-close-button-color|#000000|
|--mc-close-button-cursor|pointer|
|--mc-close-button-display|block|
|--mc-close-button-height|22px|
|--mc-close-button-margin|5px auto|
|--mc-close-button-outline|none|
|--mc-close-button-width|22px|

### Changing the Tag Name
The Multi-Checkbox component uses the `<multi-checkbox>` tag in html. In certain cases users may want, or need, to change the tag name. This can be accomplished in Javascript by subclassing the element and registering this subclass under a new tag name.

```javascript
import {MultiCheckbox} from 'multi-checkbox';

customElements.define('my-new-tag', class extends MultiCheckbox{});
```

### Contributing
1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request.


CHANGES.md
Multi-Checkbox Changelog
--------------
#### Version 0.1.11 (2023-02-06)

- Fixed css variables to expose element wide z-index.
- Add css variable to expose dropdown max-height and overflow scrolling.


#### Version 0.1.10 (2023-02-04)

- Fixed text measuring bug.


#### Version 0.1.9 (2023-02-04)

- Exposed dropdown z-index css variable.
- Implemented summary count when value width overflows display.


#### Version 0.1.8 (2021-12-08)

- Version sync to 0.1.8.


#### Version 0.1.7 (2021-12-08)

- Fixed bug clearing slot content on dynamic update.
- Added additional CSS variables.
- Fixed CSS variable default values.


#### Version 0.1.6 (2021-11-28)

- Corrected package URLs in README.md.


#### Version 0.1.5 (2021-11-28)

- First public release.


[@smontanus]: https://github.com/smontanus


multi-checkbox.js
/**
 * @file multi-checkbox.js
 * @fileOverview Multi-checkbox vanilla javascript component. Provides
 * an input with a dropdown of multiple checkboxes that are concatenated.
 * in the input when selected. See the README file for usage.
 * @version 0.1.11
 * @author Stephen Montanus <steve@stevemontanus.com>
 * @copyright ©2021 Stephen Montanus Software Engineering.
 * @license MIT
 */

/** @constant {Object} */
const template = document.createElement('template');
template.innerHTML = `
<style>
    .multi-checkbox {
        display: var(--mc-display, inline-block);
        margin: var(--mc-margin, 0);
        width: initial;
        z-index: var(--mc-z-index, 0);
    }

    .checkbox-container {
        background: var(--mc-dropdown-background, #efefef);
        border-bottom: var(--mc-border, 1px solid #000000);
        border-left: var(--mc-border, 1px solid #000000);
        border-right: var(--mc-border, 1px solid #000000);
        border-top: none;
        border-bottom-left-radius: var(--mc-border-radius, 2px);
        border-bottom-right-radius: var(--mc-border-radius, 2px);
        box-sizing: border-box;
        box-shadow: var(--mc-dropdown-box-shadow, 3px 3px 5px 1px rgba(0,0,0,0.35));
        -webkit-box-shadow: var(--mc-dropdown-box-shadow, 3px 3px 5px 1px rgba(0,0,0,0.35));
        -moz-box-shadow: var(--mc-dropdown-box-shadow, 3px 3px 5px 1px rgba(0,0,0,0.35));
        display: block;
        max-height: var(--mc-dropdown-max-height, 100px);
        position: absolute;
        text-align: var(--mc-dropdown-text-align, left);
        transform: scale(1, 0);
        transform-origin: top left;
        transition: transform 0.5s cubic-bezier(0.65, 0, 0.35, 1);
        overflow-y: scroll;
        width: var(--mc-dropdown-width, fit-content);
        z-index: var(--mc-z-index, 0);
    }

    input {
        background: var(--mc-target-background, #efefef);
        border: var(--mc-border, 1px solid #000000);
        border-bottom-left-radius: var(--mc-border-radius, 2px);
        border-top-left-radius: var(--mc-border-radius, 2px);
        color: var(--mc-target-color, #000000);
        cursor: var(--mc-target-cursor, default);
        font: var(--mc-font, 400 0.9em Arial);
        height: var(--mc-target-height, 28px);
        margin: var(--mc-target-margin, 0);
        max-height: var(--mc-target-max-height);
        min-height: var(--mc-target-min-height);
        max-width: var(--mc-target-max-width);
        min-width: var(--mc-target-min-width);
        padding: var(--mc-target-padding, 0px 0px 0px 3px);
        outline: var(--mc-target-outline, none);
        vertical-align: var(--mc-vertical-align, middle);
        width: var(--mc-target-width, 175px);
        z-index: var(--mc-z-index, 0);
    }

    #measure-test {
        left: 0;
        position: absolute;
        visibility: hidden;
        white-space: nowrap;
    }

    svg {
        vertical-align: middle;
    }

    ::slotted(ul) {
        color: var(--mc-dropdown-color, #000000);
        font: var(--mc-font, 400 0.9em Arial) !important;
        line-height: var(--mc-dropdown-line-height, 2em);
        list-style-type: none;
        margin: var(--mc-ul-margin, 5px);
        padding: var(--mc-ul-padding, 0);
        overflow-wrap: break-word;
    }

    #toggle-button {
        background: var(--mc-toggle-button-background, #ffffff);
        border: var(--mc-border, 1px solid #000000);
        border-bottom-right-radius: var(--mc-border-radius, 2px);
        border-top-right-radius: var(--mc-border-radius, 2px);
        color: var(--mc-toggle-button-color, #000000);
        cursor: var(--mc-toggle-button-cursor, pointer);
        height: var(--mc-toggle-button-height, 30px);
        margin-left: -5px;
        outline: var(--mc-toggle-button-outline, none);
        vertical-align: var(--mc-vertical-align, middle);
        width: var(--mc-toggle-button-width, 30px);
        z-index: var(--mc-z-index, 0);
    }

    #close-button {
        background: var(--mc-close-button-background, #efefef);
        border: var(--mc-close-button-border, none);
        border-radius: var(--mc-close-button-border-radius, default);
        color: var(--mc-close-button-color, #000000);
        cursor: var(--mc-close-button-cursor, pointer);
        display: var(--mc-close-button-display, block);
        height: var(--mc-close-button-height, 22px);
        margin: var(--mc-close-button-margin, 5px auto);
        outline: var(--mc-close-button-outline, none);
        width: var(--mc-close-button-width, 22px);
    }
</style>

<div class="multi-checkbox">
    <input type="text" readonly disabled />
    <span id="measure-test"></span>
    <button id="toggle-button">
        <svg x="0px" y="0px" viewBox="0 0 330 330">
            <path fill="currentColor" d="M325.607,79.393c-5.857-5.857-15.355-5.858-21.213,
            0.001l-139.39,139.393L25.607,79.393c-5.857-5.857-15.355-5.858-21.213,0.001c-5.858,
            5.858-5.858,15.355,0,21.213l150.004,150c2.813,2.813,6.628,4.393,10.606,4.393
            s7.794-1.581,10.606-4.394l149.996-150C331.465,94.749,331.465,85.251,
            325.607,79.393z"/>
        </svg>
    </button>
    <div class="checkbox-container">
        <div><slot name="check-values" /></div>
        <button id="close-button">
        <svg x="0px" y="0px" viewBox="0 0 330 330">
            <path fill="currentColor" d="M324.001,209.25L173.997,96.75c-5.334-4-12.667-4-18,
            0L6.001,209.25c-6.627,4.971-7.971,14.373-3,21c2.947,3.93,7.451,6.001,12.012,
            6.001c3.131,0,6.29-0.978,8.988-3.001L164.998,127.5l141.003,105.75c6.629,
            4.972,16.03,3.627,21-3C331.972,223.623,330.628,214.221,324.001,209.25z"/>
        </svg>
        </button>
    </div>
</div>`;

/**
 * Creates a new MultiCheckbox custom HTML element. Utilized in HTML as the
 * <multi-checkbox> tag.
 * @name MultiCheckbox
 * @class
 * @extends HTMLElement
 */
export class MultiCheckbox extends HTMLElement {
    /**
     * @description Specify observed attributes so that attributeChangedCallback will work.
     * @return {String[]} String array of attributes to be observed.
     */
    static get observedAttributes() {
        return ['separator', 'value'];
    }

    /**
     * @description Create a MultiCheckbox element. Fires when an instance of
     * the element is created or updated.
     */
    constructor() {
        super();
        this.attachShadow({mode: 'open'});
        this.shadowRoot.appendChild(template.content.cloneNode(true));
        // Component variables.
        this.dropDownVisible = false;
        this.inputRadius = this.shadowRoot.querySelector('input')
            .style.borderBottomLeftRadius;
        this.toggleRadius = this.shadowRoot.querySelector('#toggle-button')
            .style.borderBottomRightRadius;
        this.items = [];
    }

    /**
     * @description Get the MultiCheckbox element value.
     * @return {string} The value.
     */
    get value() {
        return this.getAttribute('value');
    }

    /**
     * @description Set the MultiCheckbox element value.
     * @param {string} newValue The new value of the element.
     */
    set value(newValue) {
        this.setAttribute('value', newValue);
    }

    /**
     * @description Get the MultiCheckbox item separator value.
     * @return {string} The separator value.
     */
    get separator() {
        return this.getAttribute('separator');
    }

    /**
     * @descriptionSet the MultiCheckbox item separator value.
     * @param {string} newValue The new value of the item separator.
     */
    set separator(newValue) {
        this.setAttribute('separator', newValue);
    }

    /**
     * @name updateItems
     * @description Update the item list when slot content changes.
     * @return {void}
     */
    updateItems() {
        // Remove original slot data.
        const itemList = this.shadowRoot.querySelector('slot')
            .assignedNodes({flatten: true})[0];
        if (itemList.childNodes[itemList.childNodes.length - 1].nodeName == 'UL') {
            itemList.removeChild(itemList.childNodes[itemList.childNodes.length - 1]);
        }
        // Update slot with new elements.
        this.items = this.shadowRoot.querySelector('slot')
            .assignedElements({flatten: true})[0]
            .children;
        // Add checkboxes to the component list items.
        this.addCheckBoxes();
        // Clear the input and custom element value.
        this.shadowRoot.querySelector('input').value = '';
        this.setAttribute('value', '');
    }

    /**
     * @name addCheckBoxes
     * @description Add check box inputs to the elements of the items list.
     * @return {void}
     */
    addCheckBoxes() {
        for (let i = 0; i < this.items.length; i++) {
            this.items[i].innerHTML = '<input type="checkbox" style="' +
            'height: var(--mc-checkbox-height, auto); ' +
            'vertical-align: middle; ' +
            'width: var(--mc-checkbox-width, auto);" ' +
            'id="item-' + i + '" value="' + this.items[i].innerHTML + '" /> ' +
            this.items[i].innerHTML;
        }
    }

    /**
     * @name updateValue
     * @description Update the custom element value based on checked boxes.
     * @return {void}
     */
    updateValue() {
        // Only run this when checkboxes are being clicked, not dynamically changed.
        if (this.dropDownVisible == true) {
            // Get the separator string.
            const sep = this.getAttribute('separator');
            // Initialize the string to store the new element value.
            let newVal = '';
            // Iterate over the list items, adding the values of checked checkboxes to
            // the new string.
            for (let i = 0; i < this.items.length; i++) {
                const chkBox = this.items[i].querySelector('input[type="checkbox"]');
                if (chkBox.checked) {
                    // Concatenate the checked value.
                    newVal = newVal + chkBox.value + sep;
                }
            }
            // Remove the last separator.
            newVal = newVal.slice(0, -sep.length);
            // Update the value attribute to the new string.
            this.setAttribute('value', newVal);
        }
    }

    /**
     * @name showDropDown
     * @description Show the drop down item list.
     * @return {void}
     */
    showDropDown() {
        this.shadowRoot.querySelector('.checkbox-container').style.transform = 'scale(1, 1)';
        this.shadowRoot.querySelector('input').style.borderBottomLeftRadius = '0px';
        this.shadowRoot.querySelector('#toggle-button').style.borderBottomRightRadius = '0px';
        this.shadowRoot.querySelector('#toggle-button').innerHTML =
            `<svg x="0px" y="0px" viewBox="0 0 330 330">
                <path fill="currentColor" d="M324.001,209.25L173.997,96.75c-5.334-4-12.667-4-18,
                0L6.001,209.25c-6.627,4.971-7.971,14.373-3,21c2.947,3.93,7.451,6.001,12.012,
                6.001c3.131,0,6.29-0.978,8.988-3.001L164.998,127.5l141.003,105.75c6.629,
                4.972,16.03,3.627,21-3C331.972,223.623,330.628,214.221,324.001,209.25z"/>
            </svg>`;
    }

    /**
     * @name hideDropDown
     * @description Hide the drop down item list.
     * @return {void}
     */
    hideDropDown() {
        this.shadowRoot.querySelector('.checkbox-container').style.transform = 'scale(1, 0)';
        this.shadowRoot.querySelector('input').style.borderBottomLeftRadius = this.inputRadius;
        this.shadowRoot.querySelector('#toggle-button')
            .style.borderBottomRightRadius = this.toggleRadius;
        this.shadowRoot.querySelector('#toggle-button').innerHTML =
            `<svg x="0px" y="0px" viewBox="0 0 330 330">
                <path fill="currentColor" d="M325.607,79.393c-5.857-5.857-15.355-5.858-21.213,
                0.001l-139.39,139.393L25.607,79.393c-5.857-5.857-15.355-5.858-21.213,0.001c-5.858,
                5.858-5.858,15.355,0,21.213l150.004,150c2.813,2.813,6.628,4.393,10.606,4.393
                s7.794-1.581,10.606-4.394l149.996-150C331.465,94.749,331.465,85.251,
                325.607,79.393z"/>
            </svg>`;
    }

    /**
     * @name toggleDropDown
     * @description Toggle display of the drop down item list.
     * @return {void}
     */
    toggleDropDown() {
        this.dropDownVisible = !this.dropDownVisible;
        if (this.dropDownVisible) {
            this.showDropDown();
        } else {
            this.hideDropDown();
        }
    }

    /**
     * @name measureText
     * @description Measure the element text in pixels compared to input element
     * width.
     * @return {boolean}
     */
    measureText() {
        const valuePixelWidth = this.shadowRoot.querySelector('#measure-test')
            .getBoundingClientRect().width;
        const inputPixelWidth = this.shadowRoot.querySelector('input')
            .getBoundingClientRect().width;
        if (valuePixelWidth > inputPixelWidth) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * @name connectedCallback
     * @description Fires when an instance is inserted into the document.
     * @return {void}
     */
    connectedCallback() {
        // Add button event listeners.
        this.shadowRoot.querySelector('#toggle-button')
            .addEventListener('click', () => this.toggleDropDown());
        this.shadowRoot.querySelector('#close-button')
            .addEventListener('click', () => this.toggleDropDown());
        // Add slot change event listener
        this.shadowRoot.addEventListener('slotchange', () => this.updateItems());
        // Add checkbox event listener.
        this.shadowRoot.addEventListener('click', (event) => {
            if (event.target && event.target.id.split('-')[0] == 'item') {
                this.updateValue();
            }
        });
        // Add event listener to hide the dropdown when click is outside the component.
        document.addEventListener('click', (event) => {
            if (!event.composedPath().includes(this)) {
                this.dropDownVisible = false;
                this.hideDropDown();
            }
        });
    }

    /**
     * @name disconnectedCallback
     * @description Fires when an instance was removed from the document.
     * @return {void}
     */
    disconnectedCallback() {
        // Remove button event listeners.
        this.shadowRoot.querySelector('#toggle-button')
            .removeEventListener('click', () => this.toggleDropDown());
        this.shadowRoot.querySelector('#close-button')
            .removeEventListener('click', () => this.toggleDropDown());
        // Remove slot event listener.
        this.shadowRoot.querySelector('slot')
            .removeEventListener('slotchange', () => this.updateItems());
    }

    /**
     * @name attributeChangedCallback
     * @description Fires when an attribute was added, removed, or updated.
     * @param {string} attrName The element attribute name.
     * @param {string} oldVal The old value of the attribute.
     * @param {string} newVal The new value of the attribute.
     * @return {void}
     */
    attributeChangedCallback(attrName, oldVal, newVal) {
        switch (attrName) {
        case 'value':
            // Update the text measurement hidden span element.
            this.shadowRoot.querySelector('#measure-test').textContent = newVal;
            // Update the input.
            if (this.measureText()) { // Value string is wider than input width.
                // Get the separator string.
                const sep = this.getAttribute('separator');
                // Split newVal into array using separator and get length.
                const newValCount = newVal.split(sep).length;
                this.shadowRoot.querySelector('input').value = newValCount + ' Selected';
            } else { // Value string is narrower than input width.
                this.shadowRoot.querySelector('input').value = newVal;
            }
            if (this.dropDownVisible == false) {
                // Get the separator string.
                const sep = this.getAttribute('separator');
                // Split newVal into array using separator
                const newVals = newVal.split(sep);
                // Update the checkboxes.
                for (let i = 0; i < this.items.length; i++) {
                    const chkBox = this.items[i].querySelector('input[type="checkbox"]');
                    if (chkBox) { // Filter out nulls on page load.
                        if (newVals.includes(chkBox.value)) {
                            chkBox.checked = true;
                        } else {
                            chkBox.checked = false;
                        }
                    }
                }
            }
            break;
        case 'separator':
            if (newVal !== oldVal) {
                let valueString = this.getAttribute('value');
                // Replace the separator string.
                valueString = valueString.split(oldVal).join(newVal);
                // Update the input and custom element value.
                this.shadowRoot.querySelector('input').value = valueString;
                this.setAttribute('value', valueString);
            }
            break;
        }
    }
}

customElements.define('multi-checkbox', MultiCheckbox);
